changefeedccl: Rework error handling

miretskiy opened the pull request:
Prior to this PR, changefeeds would rely on a white list approach in order to determine which errors were retryable. All other errors would be deemed terminal, causing changefeed to fail.

The above approach is brittle, and causes unwanted changefeed termination.

This PR changes this approach to treat all errors as retryable, unless otherwise indicated.  Errors that are known by changefeed to be fatal are handled explicitly, by marking such errors as terminal.  For example, changefeeds would exit
if the targetted table is dropped.  On the other hand, inability to read this table for any reason would not be treated as terminal.

Fixes #90320
Fixes #77549
Fixes #63317
Fixes #71341
Fixes #73016
Informs CRDB-6788
Informs CRDB-7581

Release note (enterprise change): Changefeed will now treat all errors, unless otherwise indicated, as retryable errors.
Release justification (stability): Improve stability of changefeeds by retrying all errors except those deemed terminal.

Commit:
changefeedccl: Rework error handling

Prior to this PR, changefeeds would rely on a white list
approach in order to determine which errors were retryable.
All other errors would be deemed terminal, causing changefeed
to fail.

The above approach is brittle, and causes unwanted
changefeed termination.

This PR changes this approach to treat all errors as retryable,
unless otherwise indicated.  Errors that are known by changefeed
to be fatal are handled explicitly, by marking such errors
as terminal.  For example, changefeeds would exit
if the targetted table is dropped.  On the other hand, inability
to read this table for any reason would not be treated as
terminal.

Fixes #90320
Fixes #77549
Fixes #63317
Fixes #71341
Fixes #73016
Informs CRDB-6788
Informs CRDB-7581

Release note (enterprise change): Changefeed will now treat
all errors, unless otherwise indicated, as retryable errors.

HonoreDB commented:
Does anything weird happen in a mixed version state? In particular, if a 22.0.0 coordinator node gets an error it would normally retry from a changefeed aggregator on a node running a version with this backport, will it look for the retryable error marker, not find it, and then sometimes erroneously treat it as a fatal error? Maybe for the backport to be safe we should keep the retryable error marking in and just not do anything with it.

(I'd be surprised if there were problems in the other direction, with a backported node not knowing what to do with a retryable error marker, but idk).

miretskiy commented:
> Does anything weird happen in a mixed version state

Interesting question; if so... it would seem to imply that master might be impacted too; and if we add 
version gate, then it's not backportable...

Commit:
changefeedccl: Fix error handling in mixed version state

Prior PR #90810 changed error handling approach used
by CDC to be "retry by default".

The above PR failed to account for some of the
situations that may arrise in mixed version state.
In particular, when upgrading, if the aggregator node
gets restarted with the new version, while the coordinator
still runs old binary, *and* the new aggregator encounters
a retryable error, that error would not be explicitly marked
as retryable, causing the old coordinator binary to treat
the error as a terminal error.

Other combination (new coordinator, old aggregator) is not
succeptible to this situation.

This PR partially reverts changes in the #90810 so previously
retryable errors continue to be explicitly marked as retryable.

There is no need to introduce version gates since the error
handling should be backward compatible, and with this PR, operate
correctly in the mixed version state.

Epic: None

Release note: none

miretskiy commented:
> Does anything weird happen in a mixed version state? In particular, if a 22.0.0 coordinator node gets an error it would normally retry from a changefeed aggregator on a node running a version with this backport, will it look for the retryable error marker, not find it, and then sometimes erroneously treat it as a fatal error? Maybe for the backport to be safe we should keep the retryable error marking in and just not do anything with it.
> 
> (I'd be surprised if there were problems in the other direction, with a backported node not knowing what to do with a retryable error marker, but idk).

Thanks for catching this; 
I added a cherry pick to partially revert this behavior.  Please take a look, @HonoreDB 
(it's possible that some tests may need to be updated).
